Add rules for `ref` safety (#742)
dotnet/csharpstandard: standard/variables.md

```

- The definite-assignment state of *v* before *expr_cond* is the same as the state of *v* before *expr*.

> - The definite-assignment state of *v* before *expr_true* is definitely assigned if the state of *v* after *expr_cond* is definitely assigned or “definitely assigned after true expression”1057.

Reviewer: I'm guessing this 1057 is an accident.

```suggestion
- The definite-assignment state of *v* before *expr_true* is definitely assigned if the state of *v* after *expr_cond* is definitely assigned or “definitely assigned after true expression”.
```

## 9.6 Atomicity of variable references

Reads and writes of the following data types shall be atomic: `bool`, `char`, `byte`, `sbyte`, `short`, `ushort`, `uint`, `int`, `float`, and reference types. In addition, reads and writes of enum types with an underlying type in the previous list shall also be atomic. Reads and writes of other types, including `long`, `ulong`, `double`, and `decimal`, as well as user-defined types, need not be atomic. Aside from the library functions designed for that purpose, there is no guarantee of atomic read-modify-write, such as in the case of increment or decrement.

## §ref-span-safety Reference variables and returns

### §ref-span-safety-general General

> A *reference variable* is a local variable or `struct` declared with the `ref` modifier. A ref local does not create a new storage location. Instead, a ref ref local represents the same storage location as its initializing expression. Thus, the value of a reference variable is always the same as the underlying variable.

Reviewer 2: I'm confused as to what a local variable and a type have in common - it would be odd to call the struct itself a variable.

Additionally, is "ref ref local" meant to be "ref local"?

> A `ref struct` may include `ref struct` or ref-like fields. A ref-like field does not create a new storage location. Instead, if refers to the same storage as its initializing expression. Ref struct types include `Span<T>`, `ReadOnlySpan<T>`, and other types that include an unmanaged pointer as a member.

Reviewer 3: > A ref-like field does not create a new storage location

This doesn't feel right to me. A ref field, or a field of ref struct type, is a storage location, and can be assigned using `v.refField = ref expr` or `v.refStructField = expr` respectively. In the ref field case, it just happens that the thing the storage location contains is a ref.

Author: Good catch. I did rework these paragraphs. still not sure it's done yet.

> A `ref struct` is a `struct` whose fields include a `ref` struct or a ref-like field. A ref-like field does not create a new storage location. Instead, if refers to the same storage as its initializing expression. Ref struct types include `Span<T>`, `ReadOnlySpan<T>`, and other types that include an unmanaged pointer as a member.

Reviewer 2: I'd expect a `ref struct` to be ` struct` declared with the `ref` modifier - with restrictions on when that modifier can be applied. (I'd expect it to be *possible* to declare a `ref struct` without any fields that are ref structs or ref-like fields, even if it would normally be pointless to do so.)

Additionally, should "`ref` struct" be "`ref struct`"? (Or just ref struct? Currently we've got three variations of using code font in this paragraph...)

Author: I made updates. On the first point, what's important here is that `ref struct` types have "ref-like" fields, so therefore have specific rules in this section.

A *reference return* is the expression returned by reference from a method whose return type includes the `ref` or `ref readonly` modifiers (§14.6.1). The return refers to the same storage as its expression.

### §ref-span-safety-escape-scopes Safe to escape scopes

> At compile-time, each expression is associated with a scope that expression is permitted to escape to, *safe-to-escape-scope*. Each variable is associated with the scope a reference to it is permitted to escape to, *ref-safe-to-escape-scope*. For a given variable expression, these may be different. When the *ref-safe-to-escape* scope is *calling-method*, the variable is *safe-to-return*. A variable that is *safe-to-return* may escape the enclosing method as a whole. In other words, the variable can *returned-by-ref*.

Reviewer 2: ```suggestion
At compile-time, each expression is associated with a scope that expression is permitted to escape to, its *safe-to-escape-scope*. Each variable is associated with the scope a reference to it is permitted to escape to, *ref-safe-to-escape-scope*. For a given variable expression, these may be different. When the *ref-safe-to-escape* scope is *calling-method*, the variable is *safe-to-return*. A variable that is *safe-to-return* may escape the enclosing method as a whole. In other words, the variable can *returned-by-ref*.
```

Or something similar - basically having *safe-to-escape-scope* on its own after the comma reads a bit oddly to me. Ditto for the second sentence. Another option would be to move the name earlier, e.g.

> At compile-time, each expression is associated with a *safe-to-escape-scope*, which is the scope that the expression is permitted to escape to.

> At compile-time, each expression is associated with a scope that expression is permitted to escape to, *safe-to-escape-scope*. Each variable is associated with the scope a reference to it is permitted to escape to, *ref-safe-to-escape-scope*. For a given variable expression, these may be different. When the *ref-safe-to-escape* scope is *calling-method*, the variable is *safe-to-return*. A variable that is *safe-to-return* may escape the enclosing method as a whole. In other words, the variable can *returned-by-ref*.

Reviewer 4: The first sentence refers to "each expression", the second refers to "each variable", and the third refers to "a given variable expression". Are the sentences referring to the same items? If so, consider using a common name - perhaps "expression", or if this is specifically about `ref` cases, perhaps "ref expression". (Are the expressions always variables or is the expression in `return ref F();` also an example?)

> When the *ref-safe-to-escape* scope is *calling-method*, the variable is *safe-to-return*. A variable that is *safe-to-return* may escape the enclosing method as a whole. In other words, the variable can *returned-by-ref*.

When the ref-safe-to-escape scope is _calling-method_, the variable is safe-to-return _by-ref_. (When the safe-to-escape scope is _calling-method_, the variable is safe-to-return _by-value_.)

Author: I need to think about this one more. This was intentional, but it's awkward. Not all expressions can be ref assigned or returned. Only variables (which have defined storage) can be assigned to a ref vaiable.

I'll keep thinking on the wording here.

> Given an assignment from an expression E1 with a *safe-to-escape-scope* S1, to a (variable) expression E2 with *safe-to-escape-scope* S2, it is an error if S2 is a wider scope than S1. By construction, the two scopes S1 and S2 are in a nesting relationship, because a legal expression is always *safe-to-return* from some scope enclosing the expression.

Reviewer 2: Do we define "wider" anywhere for scopes? (Can scopes ever overlap without nesting?)

Author: Let's discuss the terms in our meeting Wednesday.  The standard defines "scope" in S 7

> The scope of a name is the region of program text within which it is possible to refer to the entity declared by the name without qualification of the name. 

In this section, we're using "scope" to address a related to different concept: The region where the storage referred to by an "initializing expression" is valid. Maybe a different term is better? 

However, to answer your question, as defined in "Basic Concepts", scopes can be nested, but don't overlap.

Author: I updated the definitions in the general section for reference variables.

> Given an assignment from an expression E1 with a *safe-to-escape-scope* S1, to a (variable) expression E2 with *safe-to-escape-scope* S2, it is an error if S2 is a wider scope than S1. By construction, the two scopes S1 and S2 are in a nesting relationship, because a legal expression is always *safe-to-return* from some scope enclosing the expression.

Reviewer 4: > ... because a legal expression is always *safe-to-return* from some scope enclosing the expression.

I may be misunderstanding, but a `ref struct` value is not necessarily *safe-to-return* from the enclosing scope. For instance, the expression in `return s;` is not *safe-to-return*:
```csharp
static Span<int> F()
{
    Span<int> s = stackalloc int[1];
    return s; // error
}
```

Author: I think that clause can be deleted. It doesn't add anything.

Author: This is where I've added a sentence defining "wider" and "narrower" scopes.

> There are three different values for *ref-safe-to-escape-scope*: *block in which variable is declared*, *current-method* and *calling-method*. When the *ref-safe-to-escape-scope* is *calling-method*, the variable is *safe-to-return* by reference.

Reviewer 2: It feels odd to me that one of these names is space-separated, and the others are dash-separated.

Author: Fixed. I'd really like to find a shorter term for *block-in-which-variable-is-declared* though.

> There are three different values for *ref-safe-to-escape-scope*: *block in which variable is declared*, *current-method* and *calling-method*. When the *ref-safe-to-escape-scope* is *calling-method*, the variable is *safe-to-return* by reference.

Reviewer 2: Also, should calling-method actually be caller-scope or something like that? (Just thinking that the caller might not be a method...)

Author: I like caller-scope. Agreed that it could be from a property or index accessor.

An expression whose type is not a `ref struct` type is *safe-to-return* from the entire enclosing method. Otherwise we refer to the rules below.

### §ref-span-safety-locals Local variable escape scope

> For a local variable `l`:

Reviewer 2: I'd suggest using a different name here, for all the normal reasons that `l` is often hard to read or ambiguous. Perhaps `v`?

- If `l` is a `ref` variable, its *ref-safe-to-escape-scope* is taken from the *ref-safe-to-escape-scope* of its initializing expression.
- Otherwise its *ref-safe-to-escape-scope* is the scope in which it was declared.

The *safe-to-escape-scope* of an expression that designates the use of a local variable is determined as follows:

- A local whose type is not a `ref struct` type is *safe-to-return* from the entire enclosing method. Its *safe-to-escape-scope* is the *calling-method*.
- If the variable is an iteration variable of a `foreach` loop, then the variable's *safe-to-escape-scope* is the same as the *safe-to-escape-scope* of the `foreach` loop's expression.

> - A local of `ref struct` type and uninitialized at the point of declaration is *safe-to-return* from the entire enclosing method. Its *safe-to-escape-scope* is the calling method.

Reviewer 2: ```suggestion
- A local of `ref struct` type which is uninitialized at the point of declaration is *safe-to-return* from the entire enclosing method. Its *safe-to-escape-scope* is the *calling method*.
```

(Or possible *calling-method*. We need to work out how we want to deal with these, then be consistent.)

> - Otherwise the variable's type is a `ref struct` type, and the variable's declaration requires an initializer. The variable's *safe-to-escape-scope* is the same as the *safe-to-escape* of its initializer.

Reviewer 2: Requires an initializer, or has an initializer?

Author: has is probably better.

> ### §ref-span-safety-parameters Parameter escape scope

Author: This section doesn't handle a rule needed later in C# 11 (when `ref` fields are allowed):

-  A `ref` parameter can't be returned by `ref` as a field in another `ref` argument. 

For example:

```csharp
ref struct S { 
  ref int refField;
}

void M(ref S p1, ref int p2) {
   p1.refField = ref p2;
}
```

These rules would allow that because both `p1` and `p2` have *ref-safe-to-escape-scope* of *calling-method*. But, doing so could be dangerous:  `p2` could be a local in the calling method, and `p1` could be a static field in the type of the calling method.

For a formal parameter `p`:

- If `p` is a `ref`, or `in` parameter, its *ref-safe-to-escape-scope* is the *calling-method*. It is *safe-to-return* by ref. If `p` is an `in` parameter, it can't be returned as a writable `ref` but can be returned as `ref readonly`.
- If `p` is an `out` parameter, its *ref-safe-to-escape-scope* is the *current-method*. It isn't *safe-to-return* by ref.
- Otherwise, if `p` is the `this` parameter of a struct type, its *ref-safe-to-escape-scope* is the *current-method*.
- Otherwise, the parameter is a value parameter, and it is *ref-safe-to-escape-scope* is the *current-method*

For an expression that designates the value of a formal parameter `p`, its *safe-to-escape-scope* is the calling method. This applies to the `this` parameter.

### §ref-span-safety-field-reference Field reference escape scope

For a variable designating a reference to a field, `e.F`:

- If `e` is of a reference type, its *ref-safe-to-escape-scope* is the *calling-method*.

> - Otherwise, if `e` is of a value type, its *ref-safe-to-escape-scope* is taken from the *ref-safe-to-escape* of `e`.

Reviewer 4: Should this section also include ref fields? If the field `e.F` is a ref field, where `e` is a of `ref struct` type, the _ref-safe-to-escape-scope_ of `ref e.F` is the _safe-to-escape_ of `e`.

Author: Not yet. The committee is on V7.x, so ref fields aren't supported yet.

For an expression that designates a reference to a field, `e.F`, its a *safe-to-escape-scope* is the same as the *safe-to-escape-scope* of `e`.

### §ref-span-safety-operators Operators including `?:`

The application of a user-defined operator is treated as a method invocation.

> For an operator that yields an value, such as `e1 + e2` or `c ? e1 : e2`, the *safe-to-escape-scope* of the result is the narrowest scope among the *safe-to-escape-scopes* of the operands of the operator.  As a consequence, for a unary operator that yields an value, such as `+e`, the *safe-to-escape-scope* of the result is the *safe-to-escape-scope* of the operand.

Reviewer 2: ```suggestion
For an operator that yields a value, such as `e1 + e2` or `c ? e1 : e2`, the *safe-to-escape-scope* of the result is the narrowest scope among the *safe-to-escape-scopes* of the operands of the operator.  As a consequence, for a unary operator that yields an value, such as `+e`, the *safe-to-escape-scope* of the result is the *safe-to-escape-scope* of the operand.
```

And do we formally define "narrowest scope" anywhere? (We may decide we don't need to.)

> For an operator that yields an variable, such as `c ? ref e1 : ref e2`:

Reviewer 2: ```suggestion
For an operator that yields a variable, such as `c ? ref e1 : ref e2`:
```

- The *ref-safe-to-escape-scope* of the result is the narrowest scope among the *ref-safe-to-escape-scopes* of the operands of the operator.
- The *safe-to-escape-scope* of the operands must agree, and that is the *safe-to-escape-scope* of the resulting variable.

### §ref-span-safety-method-invocation Method invocation

> For a variable `c` resulting from a ref-returning method invocation, `ref e1.M(e2, ...)`, its *ref-safe-to-escape-scope* is the smallest of the following scopes:

Reviewer 2: Narrowest instead of smallest, for consistency?

- The *calling-method*.
- The *ref-safe-to-escape-scope* of all `ref` and `out` argument expressions (excluding the receiver).
- For each `in` parameter of the method, if there is a corresponding expression that is an variable, its *ref-safe-to-escape-scope*, otherwise the nearest enclosing scope
- The *safe-to-escape-scope* of all argument expressions (including the receiver)

> *Example*: the last bullet is necessary to handle code such as
>
> ```csharp
> var sp = new Span(...)
> return ref sp[0];
> ```
>
> or
>
> ```csharp
> return ref M(sp, 0);
> ```
>
> *end example*

> For a value resulting from a method invocation `e1.M(e2, ...)`, its *safe-to-escape-scope* from the smallest of the following scopes:

Reviewer 2: Narrowest?

- The *calling-method*
- The *safe-to-escape* of all argument expressions (including the receiver)

> A property invocation (either `get` or `set`) it treated as a method invocation of the underlying method by the above rules.

Reviewer 2: ```suggestion
A property invocation (either `get` or `set`) is treated as a method invocation of the underlying method by the above rules.
```

### §ref-span-safety-a-value A value

A value's *ref-safe-to-escape-scope* is the nearest enclosing scope.

> *Note:* This occurs in an invocation such as `M(ref d.Length)` where `d` is of type `dynamic`. It is also consistent with arguments corresponding to `in` parameters.

### §ref-span-safety-stackalloc `stackalloc`

> For a value `c` resulting from a `stackalloc` expression its *safe-to-escape-scope* to the current method.

Reviewer 2: ```suggestion
The *safe-to-escape-scope* for a value `c` resulting from a `stackalloc` expression is the current method.
```

#### §ref-span-safety-constructor-invocations Constructor invocations

A `new` expression that invokes a constructor obeys the same rules as a method invocation (§ref-span-safety-method-invocation) that is considered to return the type being constructed.

In addition, for a value `c` that is the result of a `new` expression, the *safe-to-escape-scope* is no wider than the smallest of the *safe-to-escape-scopes* of all arguments and operands of the object initializer expressions, recursively, if any initializer is present.

### §ref-span-safety-default-expressions `default` expressions

> For a `default` expression, the *safe-to-escape-scope* is the *calling-method*.

Reviewer 2: ```suggestion
The *safe-to-escape-scope* of a `default` expression is the *calling-method*.
```

(This and the previous suggestion just feel like more straightforward statements.)

### §ref-span-safety-limitations Limitations on reference variables

> - Neither a `ref` parameter, nor a `ref` local, nor a parameter or local of a `ref struct` type can be lifted into a lambda or local function.

Reviewer 2: ```suggestion
- Neither a `ref` parameter, nor a `ref` local, nor a parameter or local of a `ref struct` type can be lifted into a lambda expression or local function.
```

> - Neither a `ref` parameter nor a parameter of a `ref struct` type may be an argument on an iterator method or an `async` method.

Reviewer 2: ```suggestion
- Neither a `ref` parameter nor a parameter of a `ref struct` type may be an argument for an iterator method or an `async` method.
```

(Or perhaps "provided to"?)

## 9.6 Atomicity of variable references

Reads and writes of the following data types shall be atomic: `bool`, `char`, `byte`, `sbyte`, `short`, `ushort`, `uint`, `int`, `float`, and reference types. In addition, reads and writes of enum types with an underlying type in the previous list shall also be atomic. Reads and writes of other types, including `long`, `ulong`, `double`, and `decimal`, as well as user-defined types, need not be atomic. Aside from the library functions designed for that purpose, there is no guarantee of atomic read-modify-write, such as in the case of increment or decrement.

## §ref-span-safety Reference variables and returns

### §ref-span-safety-general General

> A *reference variable* is a local variable declared with the `ref` modifier, or an instance of a `ref struct` type. A ref local does not create a new storage location. Instead, a ref local represents the same storage location as its initializing expression. Thus, the value of a reference variable is always the same as the underlying variable.

Reviewer: A, say, `int` local variable declaration creates a storage location of the bits of an `int` value; if  a `ref int` is declared where are the bits of the `ref int` value stored? C++ says “*It is unspecified whether or not a reference requires storage*”, are we saying in C# a reference variable *never* requires storage?

A `ref struct` may include `ref struct` or ref-like fields. A ref-like field refers to the same storage as its initializing expression. Its storage size in the ref struct is the same storage size of a reference field. Unlike a reference field, a ref-like field may refer to a `struct` whose storage may be on the execution stack. Ref struct types include `Span<T>`, `ReadOnlySpan<T>`, and other types that include an unmanaged pointer as a member.

A *reference return* is the expression returned by reference from a method whose return type includes the `ref` or `ref readonly` modifiers (§14.6.1). The return refers to the same storage as its expression.

All reference variables must obey safety rules that ensure the lifetime of the reference variable is not greater than the lifetime of the storage it refers to.

### §ref-span-safety-escape-scopes Safe to escape scopes

> At compile-time, each expression is associated with a scope that expression is permitted to escape to, its *safe-to-escape-scope*. Each variable is associated with the scope a reference to it is permitted to escape to, *ref-safe-to-escape-scope*. For a given variable expression, these may be different. When the *ref-safe-to-escape-scope* scope is *caller-scope*, the variable is *safe-to-return*. A variable that is *safe-to-return* may escape the enclosing method as a whole. In other words, the variable can *returned-by-ref*.

Reviewer: Does the *expression* escape or does it’s value/result escape?

Author: It should be variable (for ref safe to escape) or value (for safe to escape). Values only matter if they are of a type that has a ref-like field.

Given an assignment from an expression E1 with a *safe-to-escape-scope* S1, to a (variable) expression E2 with *safe-to-escape-scope* S2, it is an error if S2 is a wider scope than S1. By construction, the two scopes S1 and S2 are in a nesting relationship.

There are three different values for *ref-safe-to-escape-scope*: *block-in-which-variable-is-declared*, *current-method* and *caller-scope*. When the *ref-safe-to-escape-scope* is *caller-scope*, the variable is *safe-to-return* by reference. These three values form a nesting relationship from narrowest (*block-in-which-variable-is-declared*) to widest (*caller-scope*).

> An expression whose type is not a `ref struct` type is *safe-to-return* from the entire enclosing method. Otherwise we refer to the rules below.

Reviewer: As previous comment, what exactly is escaping?

### §ref-span-safety-locals Local variable escape scope

For a local variable `l`:

> - If `v` is a `ref` variable, its *ref-safe-to-escape-scope* is taken from the *ref-safe-to-escape-scope* of its initializing expression.

Reviewer: Typo maybe, `v` here, `l` in previous line?

- Otherwise its *ref-safe-to-escape-scope* is the scope in which it was declared.

The *safe-to-escape-scope* of an expression that designates the use of a local variable is determined as follows:

> - A local whose type is not a `ref struct` type is *safe-to-return* from the entire enclosing method. Its *safe-to-escape-scope* is the *caller-scope*.

Reviewer: Surely a local *variable* is never *safe-to-return*, it is the contents/value of that local which is. Given we are explicitly talking about *references to variables/locations* vs. *contents/value of variables/locations" we need to be careful of the language.

Author: This should be value.

## 9.6 Atomicity of variable references

Reads and writes of the following data types shall be atomic: `bool`, `char`, `byte`, `sbyte`, `short`, `ushort`, `uint`, `int`, `float`, and reference types. In addition, reads and writes of enum types with an underlying type in the previous list shall also be atomic. Reads and writes of other types, including `long`, `ulong`, `double`, and `decimal`, as well as user-defined types, need not be atomic. Aside from the library functions designed for that purpose, there is no guarantee of atomic read-modify-write, such as in the case of increment or decrement.

## §ref-span-safety Reference variables and returns

### §ref-span-safety-general General

> A *reference variable* is a variable that refers to another variable, called the *referent*. Therefore, a *reference variable* does not store its value. Instead, a *reference variable* refers to its *referent*. Thus, the value of a reference variable is always the same as its referent variable. A *reference variable* is a local variable declared with the `ref` modifier, or an instance of a `ref struct` type.

Reviewer 2: I'm not sure what's meant by "or an instance of a `ref struct` type" here. Is an instance of a `ref struct` type always a variable? Is the point that if we call a method that returns a `ref struct`, we can treat that return value as a reference variable? An example would at least help me understand this - not sure yet whether it needs to be in the standard though!

Author: This is an area where the standard language gets much easier in C# 11. In C# 11, a `ref struct` can contain `ref` fields. Therefore, returning a `ref struct` by value returns the `ref` field, which is a reference.

In C# 7.n (up to C# 11), types such as `Span<T>` have *ref-like* fields. They are effectively references, but the compiler doesn't allow it yet, so it's a special case.

Reviewer 2: Hmm. It's still unclear to me how the end of this sentence works. We may be able to leave it as a fudge for now though.

Author: I added three examples in this section: one for a local `ref` variable. One for an instance of a `ref struct`, and one for a `return ref`. I think the `ref struct` sample should address the "fudginess".

> A *reference variable* is a variable that refers to another variable, called the *referent*. Therefore, a *reference variable* does not store its value. Instead, a *reference variable* refers to its *referent*. Thus, the value of a reference variable is always the same as its referent variable. A *reference variable* is a local variable declared with the `ref` modifier, or an instance of a `ref struct` type.

Reviewer 2: Should some of these italicised terms be bold as well (as we're defining them)? @RexJaeschke to comment...

> A `ref struct` may include `ref struct` or ref-like fields. A ref-like field refers to the same storage as its initializing variable. Its storage size in the ref struct is the same storage size of a reference field. Unlike a reference field, a ref-like field may refer to a `struct` whose storage may be on the execution stack. Ref struct types include `Span<T>`, `ReadOnlySpan<T>`, and other types that include an unmanaged pointer as a member.

Reviewer 2: Ref-like in bold and italics where it's defined? (It's not clear whether the second sentence is meant to be a definition or not.)

Author: I intended the second sentence to be the definition. @RexJaeschke to comment on style.

> A *reference return* is the expression returned by reference from a method whose return type includes the `ref` or `ref readonly` modifiers (§14.6.1). The variable expression of the *reference return* is the *referent* of the *reference return*.

Reviewer 2: Definitely worth an example here (eventually; doesn't need to be in this PR).

Author: Should I break up or move the examples starting on line 934? If so, I'll make those edits in this PR.

Reviewer 2: I think just a single example of this would be useful - just to anchor the terms "reference return" and "referent" in something concrete.

Author: Three examples added.

> All reference variables obey safety rules that ensure the valid scope of the reference variable is not greater than the valid scope of its referent.

Reviewer 2: Have we defined what "valid scope" means? We definitely know what a scope is, but I'm not sure whether valid scope is different.

Author: Re-wrote this sentence and moved to the first paragraph in the next section.

> ### §ref-span-safety-escape-scopes Safe scopes

Reviewer 2: Perhaps call this "Safe scopes and ref-safe scopes"?

Author: For the next commit, I went with "Ref safe scopes".

Leaving this unresolved because of the question I have on avoiding *safe_scope* as a normative term.

For any variable, the *ref_safe_scope* of that variable is the scope where a *variable_reference* (§9.5) to that variable is valid. The *referent* of a reference variable must have a *ref_safe_scope* that is at least as wide as the scope where the reference variable is valid.

There are three valid *ref_safe_scope*s:

> - *block*: A variable declared in a block is valid in that block in which it is declared. This includes all local variables. A local variable declared in a method is valid in the block that defines the method. A variable declared in a block is a valid referent only if the reference variable is declared in the same block, or a nested block.

Reviewer 2: ```suggestion
- *block*: A variable declared in a block is valid in the block in which it is declared. This includes all local variables. A local variable declared in a method is valid in the block that defines the method. A variable declared in a block is a valid referent only if the reference variable is declared in the same block, or a nested block.
```

> - *block*: A variable declared in a block is valid in the block in which it is declared. This includes all local variables. A local variable declared in a method is valid in the block that defines the method. A variable declared in a block is a valid referent only if the reference variable is declared in the same block, or a nested block.

Reviewer 2: This defines whether variables themselves are valid... which seems a bit weird. Should this be:

> A *variable_reference* to a local variable declared in a block is valid in the block in which it is declared.

Likewise for the third sentence. The last sentence looks okay already - we could rephrased all of them just by changing "is valid" to "is a valid referent" potentially.

Author: Updated to the suggested wording.

> - *method*: A value parameter on a method declaration, including the implicit `this` parameter is valid in the entire method. The fields of a `struct` type are valid in the entire method. A variable with *ref_safe_scope* of *method* is a valid referent only if the reference variable is declared in the same method.

Reviewer 2: Does this include constructors? (What about constructor initializers?) Are indexers relevant here, given that they have parameters too?

Author: Yes to all these. Also, the accessors for properties. What term would be better?

Reviewer 2: Function member?

Author: Based on this edit, I changed the term from *method* to *function member*.

> - *caller_scope*: Member fields of a `class` type and `ref` parameters have *ref_safe_scope* of *caller_scope*. A variable with *ref_safe_scope* of *caller_scope* can be the referent of a reference return. A variable that can be referent of a *reference-return* is *safe-to-ref-return*.

Reviewer 2: ```suggestion
- *caller_scope*: Member fields of a `class` type and `ref` parameters have *ref_safe_scope* of *caller_scope*. A variable with *ref_safe_scope* of *caller_scope* can be the referent of a reference return. A variable that can be the referent of a *reference-return* is *safe-to-ref-return*.
```

These values form a nesting relationship from narrowest (*block*) to widest (*caller-scope*).

> For any variable, the *safe_scope* of that variable is the scope where its value may be copied. A variable whose type is not a `ref struct` type is *safe-to-return* from the entire enclosing method. Its *safe_scope* is *caller_method*. Otherwise the following rules apply.

Reviewer 2: Should *caller_method* here be *caller_scope*? This is the only occurrence of *caller_method*.

> For any variable, the *safe_scope* of that variable is the scope where its value may be copied. A variable whose type is not a `ref struct` type is *safe-to-return* from the entire enclosing method. Its *safe_scope* is *caller_method*. Otherwise the following rules apply.

Reviewer 2: I think it would be worth moving the definition of safe_scope to be nearer the definition of ref_safe_scope. Maybe.

Author: Leaving unresolved. See my other comment on *safe_scope*

> *Example*: The following code shows examples of the different *safe_scope* and *ref_safe_scope* values. The declarations show values of *safe_scope* for a referent to be the initializing expression for a `ref` variable. The examples show the values of *ref_safe_scope* for a reference return:
>
> ```csharp
> public class C
> {
>     private int[] arr = { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9 };
> 

> >     public ref int M1(ref int rv)

Reviewer 2: I'd suggest avoid duplicating variable names. We've got `rv` in three different places (two parameters and a local variable).

>     {
>         return ref rv; // rv is safe to ref return
>     }
>
>     public ref int M2(int v)
>     {
>         return ref v; // error: v isn't safe to ref return
>     }
>
>     public ref int M3(ref int rv)
>     {
>         int v = 5;
> 
>         return ref arr[v]; // arr[v] is safe to ref return
>     }
> 
>     public void M4(int p) 
>     {
>         int v = 6;
>         ref int rv = ref p; // safe scope of rv is block, ref safe scope of p is method
>         ref int rv2 = ref v; // safe scope of rv2 is block, ref safe scope of v is block
>         ref int rv3 = ref arr[v]; // safe scope of rv3 is block, ref safe scope of arr[v] is caller method
>     }
> }
> ```
>
> *end example.*

> ### §ref-span-safety-locals Local variable safe scope

Reviewer 2: Do we have sub-heading 4 sections? If so, I think this probably should be one - otherwise "Local variable safe scope" is a peer of "Safe scopes" when I'd expect it to be a child.

Reviewer 2: (Or possibly this just shouldn't be a new section?)

Author: These should all have been H4. Fixed

> For a local variable `v`:

Reviewer 2: I was surprised to see the rule for determining a ref_safe_scope in a section about safe scopes

Author: Aha.  This gives me an idea:

The concept of *safe_scope* only is interesting with respect to `ref struct` types. For all other types, you can obviously pass by value and return by value. A `ref struct` is interesting ***because*** it's *safe_scope* is constrained by the *ref_safe_scope* of any member *ref-like* fields.

After I make the other edits. I'd like your thoughts on the following:

- Don't introduce *safe_scope* as a term.
- Instead, add a rule that any instance of a `ref struct` is constrained to the *ref_safe_scope* of the referent in any *ref_like_fields*.

Reviewer 2: I can't validate that that'll work, but I *do* like the idea of only having one term here. I'm finding the two confusing.

Reviewer 5: Using one term should work just fine as long as we are consistent. Considered doing this once or twice in the ref safety docs when we implemented `ref` fields but was too much work to go back and change the original spec.

Slight tangent ... if you think about `ref` as a type vs. an annotation on a location it often makes the rules around ref safety easier to grok (at least to me). Consider that if we had `ref<int>` instead of `ref int` then having a single term becomes completely natural here as everything is just a value.

Reviewer 2: Thanks @jaredpar - that's really interesting. On my next pass through this, I'll try reading it from that perspective. @BillWagner if that turns out to be positive, *maybe* we could add a note? It's not the kind of note we'd often add, but given the complexities of this, it might be worth it here.

> - If `v` is a `ref` variable, its *ref-safe-scope* is taken from the *ref-safe-scope* of its initializing expression.

Reviewer 2: This section, and the ones below, use hyphens instead of underscores, which I assume is a mistake?

Author: Fixed. Also, swept for the term changes.

- Otherwise its *ref-safe-scope* is the scope in which it was declared.

The *safe-scope* of an expression that designates the use of a local variable is determined as follows:

- A value whose type is not a `ref struct` type is *safe-to-return* from the entire enclosing method. Its *safe-scope* is the *caller-scope*.
- If the variable is an iteration variable of a `foreach` loop, then the variable's *safe-scope* is the same as the *safe-scope* of the `foreach` loop's expression.
- A local of `ref struct` type which is uninitialized at the point of declaration is *safe-to-return* from the entire enclosing method. Its *safe-scope* is the *calling method*.
- Otherwise the variable's type is a `ref struct` type, and the variable's declaration has an initializer. The variable's *safe-scope* is the same as the *ref-safe-scope* of its initializer.

### §ref-span-safety-parameters Parameter escape scope

For a formal parameter `p`:

- If `p` is a `ref`, or `in` parameter, its *ref-safe-scope* is the *caller-scope*. It is *safe-to-return* by ref. If `p` is an `in` parameter, it can't be returned as a writable `ref` but can be returned as `ref readonly`.

> - If `p` is an `out` parameter, its *ref-safe-to-escape-scope* is the *current-method*. It isn't *safe-to-return* by ref.

Reviewer 2: I think this is the first time ref-safe-to-escape has been mentioned, and I see various other references to safe-to-escape-scope which hasn't been defined.

If we could make sure that every term has a definition in bold and italic, it'll be easier to make sure we never use anything that's not defined.

Author: Thanks! Updated all references in this section to use the newer proposed term

- Otherwise, if `p` is the `this` parameter of a struct type, its *ref-safe-to-escape-scope* is the *current-method*.
- Otherwise, the parameter is a value parameter, and it is *ref-safe-to-escape-scope* is the *current-method*

For an expression that designates the value of a formal parameter `p`, its *safe-scope* is the calling method. This applies to the `this` parameter.

> ### §ref-span-safety-field-reference Field reference escape scope

Reviewer 2: Should this be "ref safe scope" rather than "reference escape scope"?

For a variable designating a reference to a field, `e.F`:

- If `e` is of a reference type, its *ref_safe_scope* is the *caller_scope*.
- Otherwise, if `e` is of a value type, its *ref_safe_scope* is taken from the *ref_safe_scope* of `e`.

For an expression that designates a reference to a field, `e.F`, its a *safe_scope* is the same as the *safe_scope* of `e`.

#### §ref-span-safety-operators Operators

The application of a user-defined operator is treated as a method invocation.

For an operator that yields a value, such as `e1 + e2` or `c ? e1 : e2`, the *safe_scope* of the result is the narrowest scope among the *safe_scopes* of the operands of the operator.  As a consequence, for a unary operator that yields an value, such as `+e`, the *safe_scope* of the result is the *safe_scope* of the operand.

> For an operator that yields a variable, such as `c ? ref e1 : ref e2`:

Reviewer 2: When referring to the operands here, I suspect it's only the second and third operands that are relevant, right? It doesn't matter what the safe_scope or ref_safe_scope of `c` is, does it?

Reviewer 2: (This is still a live comment.)

Author: I made a change, but I'm still not thrilled. Suggestions welcome.

Reviewer 2: In meeting: are the only relevant operators `?:` and `= ref`? In which case, let's just describe those directly.

### §ref-span-safety-a-value Values

A value's *ref-safe-scope* is the nearest enclosing scope.

> *Note:* This occurs in an invocation such as `M(ref d.Length)` where `d` is of type `dynamic`. It is also consistent with arguments corresponding to `in` parameters.

### §ref-span-safety-stackalloc stackalloc

The *safe-scope* for a value `c` resulting from a `stackalloc` expression is the current method.

#### §ref-span-safety-constructor-invocations Constructor invocations

A `new` expression that invokes a constructor obeys the same rules as a method invocation (§ref-span-safety-method-invocation) that is considered to return the type being constructed.

In addition, for a value `c` that is the result of a `new` expression, the *safe-scope* is no wider than the smallest of the *safe-to-escape-scopes* of all arguments and operands of the object initializer expressions, recursively, if any initializer is present.

### §ref-span-safety-default-expressions Default expressions

The *safe-scope* of a `default` expression is the *caller-scope*.

### §ref-span-safety-limitations Limitations on reference variables

> - Neither a `ref` parameter, nor a `ref` local, nor a parameter or local of a `ref struct` type can be lifted into a lambda expression or local function.

Reviewer 2: Reference for lifting? (Is this the same lifting as in 10.6.2 and 11.4.8?)

Author: That's an incorrect term. It should be "captured by"

> - Neither a `ref` parameter nor a parameter of a `ref struct` type may be an argument for an iterator method or an `async` method.

Reviewer 2: I think we typically use "shall" instead of "may", but I defer to @RexJaeschke on that.

Author: Defer to @RexJaeschke because this is worded as a negative, so it's mandating a compiler error. 

Rex, should these bullets all be re-written in the form "... emits a compiler error"?

Reviewer 2: Action: @BillWagner to fix both this and the line above to be unambiguous.

- Neither a `ref` local, nor a local of a `ref struct` type may be in scope at the point of a `yield return` statement or an `await` expression.
- A `ref struct` type may not be used as a type argument, or as an element type in a tuple type.
- A `ref struct` type may not be the declared type of a field, except that it may be the declared type of an instance field of another `ref struct`.
- A `ref struct` type may not be the element type of an array.

> - A value of a `ref struct` type may not be boxed:

Reviewer 2: We probably need to modify 10.2.9 for this. (It may be in your follow-up list already.) Various other restrictions below probably need propagating as well.

Author: I'd like to defer these references to these PRs:

- #44 out variables (which may just be closed now)
- #213 ref locals and returns
- #219 `in` parameters
- #601 `ref struct` types

I think the text should be in the section that defines each of those variable types. These rules may be removed at that point.

Reviewer 2: Right, that makes sense. (Although we could potentially keep these rules here as a note with references - it's handy to have them all in one place.)

Reviewer 2: Action from 2023-04-19: agree to defer.

>     {
>         // element is a reference variable that refers to arr[5]
>         ref int element = ref arr[5];
>         return ref element; // return reference to arr[5];
>     }     
> }
> ```
>
> *end example*

### §ref-span-safety-escape-scopes Ref safe scopes

All reference variables obey safety rules that ensure the scope of the reference variable is not greater than the ref-safe-scope of its referent.

For any variable, the ***ref-safe-scope*** of that variable is the scope where a *variable_reference* (§9.5) to that variable is valid. The referent of a reference variable must have a ref-safe-scope that is at least as wide as the scope of the reference variable.

There are three valid ref-safe-scopes:

- ***block***: A *variable_reference* to a local variable declared in a block is valid from its declaration to the end of the block in which it is declared. The ref-safe-scope of a local variable is block. A local variable declared in a method has a ref-safe-scope of the block that defines the method. A variable declared in a block is a valid referent only if the reference variable is declared in the same block after the referent, or a nested block.

> - ***function-member***: A *variable_reference* to a value parameter on a function member declaration, including the implicit `this` parameter is valid in the entire function member. The ref-safe-scope of the fields of a `struct` type is function-member. A variable with ref-safe-scope of function-member is a valid referent only if the reference variable is declared in the same function member.

Reviewer 2: ```suggestion
- ***function-member***: A *variable_reference* to a value parameter on a function member declaration, including the implicit `this` parameter, is valid in the entire function member. The ref-safe-scope of the fields of a `struct` type is function-member. A variable with ref-safe-scope of function-member is a valid referent only if the reference variable is declared in the same function member.
```

(Just a comma after "`this` parameter".)

- ***caller-scope***: Member fields of a `class` type and `ref` parameters have ref-safe-scope of caller-scope. A variable with ref-safe-scope of caller-scope can be the referent of a reference return. A variable that can be the referent of a reference-return is ***safe-to-ref-return***.

> These values form a nesting relationship from narrowest (*block*) to widest (caller-scope).

Reviewer 2: ```suggestion
These values form a nesting relationship from narrowest (block) to widest (caller-scope).
```

> *Example*: The following code shows examples of the different ref-safe-scopes. The declarations show the ref-safe-scope for a referent to be the initializing expression for a `ref` variable. The examples show the ref-safe-scope for a reference return:
>
> ```csharp

> > public class C

Reviewer 2: Possibly add the ref-safe-scope of `r1`, `v1`, `arr` and `arr[v2]`?

- Otherwise, if `p` is the `this` parameter of a struct type, its ref-safe-scope is the function-member.
- Otherwise, the parameter is a value parameter, and its ref-safe-scope is the function-member. It isn't safe-to-ref-return.

#### §ref-span-safety-field-reference Field ref safe scope

For a variable designating a reference to a field, `e.F`:

- If `e` is of a reference type, its ref-safe-scope is the caller-scope.
- Otherwise, if `e` is of a value type, its ref-safe-scope is the same as the ref-safe-scope of `e`.

#### §ref-span-safety-operators Operators

The conditional operator (§12.18), `c ? ref e1 : ref e2`, and reference assignment operator, `= ref e` (§12.21.1) have reference variables as operands and yield a reference variable. For those operators, the ref-safe-scope of the result is the narrowest scope among the ref-safe-scopes of all `ref` operands.

#### §ref-span-safety-method-invocation Method invocation

For a variable `c` resulting from a ref-returning method invocation, `ref e1.M(e2, ...)`, its ref-safe-scope is the narrowest of the following scopes:

- The caller-scope.
- The ref-safe-scope of all `ref` and `out` argument expressions (excluding the receiver).

> - For each `in` parameter of the method, if there is a corresponding expression that is an variable, its ref-safe-scope, otherwise the nearest enclosing scope

Reviewer 2: ```suggestion
- For each `in` parameter of the method, if there is a corresponding expression that is a variable, its ref-safe-scope, otherwise the nearest enclosing scope
```

- The scope of all argument expressions (including the receiver).

> > *Example*: the last bullet is necessary to handle code such as

Reviewer 2: It would be good to improve this example over time - I don't know what "handling" these situations really means.

Author: I replaced it with a more obvious example.

## 9.6 Atomicity of variable references

Reads and writes of the following data types shall be atomic: `bool`, `char`, `byte`, `sbyte`, `short`, `ushort`, `uint`, `int`, `float`, and reference types. In addition, reads and writes of enum types with an underlying type in the previous list shall also be atomic. Reads and writes of other types, including `long`, `ulong`, `double`, and `decimal`, as well as user-defined types, need not be atomic. Aside from the library functions designed for that purpose, there is no guarantee of atomic read-modify-write, such as in the case of increment or decrement.

## §ref-span-safety Reference variables and returns

### §ref-span-safety-general General

> A ***reference variable*** is a variable that refers to another variable, called the ***referent***. Therefore, a reference variable does not store its value. Instead, a reference variable refers to its referent. Thus, the value of a reference variable is always the same as its referent variable. A reference variable is a local variable declared with the `ref` modifier.

Reviewer: This is a bit torturous, *"a reference variable does not store its value"* – it does, what it doesn't store is the *referent's value*. I understand what is being attempted – to not talk about the *value* stored in the ref variable as much as possible, except we have to as we can reassign that value so the variable refers to a different referent…

There is a comment below by @jaredpar, its hidden now so here's an extract: 

> Slight tangent ... if you think about ref as a type vs. an annotation on a location it often makes the rules around ref safety easier to grok (at least to me). Consider that if we had ref<int> instead of ref int then having a single term becomes completely natural here as everything is just a value.

What we're being fuzzy over is that it is a type, a reference variable holds a value of type "*reference to a variable*" – echoes of Algol-68 `ref ref int x := ref int y` ;-), but that languages attempt to make all this clear confused a *lot* of people…

Anyway I'll attempt a bit of wordsmithing here but really we need to consider how we're futzing this in general:

```suggestion
A ***reference variable*** is a variable that refers to another variable, called the ***referent***. A reference variable does not store the value of its referent. When a reference variable is used where a value is required its referent's value is returned; similarly when a reference variable is the target of an assignment it is the referent which is assigned to. The variable to which a reference variable refers, i.e. its referent, can be changed using a ref assignment (`= ref`) [insert x-ref]. A reference variable is a local variable declared with the `ref` modifier.
```
I can't say I'm totally happy with that, maybe someone else can improve it, but I think its a bit clearer.

> *Example:* The following example demonstrates a local reference variable whose referent is an element of an array:
>
> ```csharp
> public class C
> {
>
>     public void M()
>     {
>         int[] arr = new int[10];
>         // element is a reference variable that refers to arr[5]
>         ref int element = ref arr[5];
>         element += 5; // arr[5] has been incremented by 5
>     }     
> }
> ```
>
> *end example*

A ***reference return*** is the expression returned by reference from a method whose return type includes the `ref` or `ref readonly` modifiers (§15.6.1). The variable expression of the reference return is the referent of the reference return.

> *Example:* The following example demonstrates a reference return whose referent is an element of an array field:
>
> ```csharp
> public class C
> {
>     private int[] arr = new int[10];
>
>     public readonly ref int M()
>     {
>         // element is a reference variable that refers to arr[5]
>         ref int element = ref arr[5];
>         return ref element; // return reference to arr[5];
>     }     
> }
> ```
>
> *end example*

> ### §ref-span-safety-escape-scopes Ref safe scopes

Author: These should not be called "scopes". 

Resolved at the meeting: Refer to these as the safe "context".  Add a *note* to explain that the compiler is inferring the valid lifetime of the variables based on the program text. The compiler determines that the lifetime would extend at least as wide as the program text of the safe context.

All reference variables obey safety rules that ensure the scope of the reference variable is not greater than the ref-safe-scope of its referent.

For any variable, the ***ref-safe-scope*** of that variable is the scope where a *variable_reference* (§9.5) to that variable is valid. The referent of a reference variable must have a ref-safe-scope that is at least as wide as the scope of the reference variable.

There are three valid ref-safe-scopes:

> - ***block***: A *variable_reference* to a local variable declared in a block is valid from its declaration to the end of the block in which it is declared. The ref-safe-scope of a local variable is block. A local variable declared in a method has a ref-safe-scope of the block that defines the method. A variable declared in a block is a valid referent only if the reference variable is declared in the same block after the referent, or a nested block.

Author: Add an explanation that blocks are nested, the same way that scopes are nested. There is a different context for each block.

Author: addressed in next commit.

- ***function-member***: A *variable_reference* to a value parameter on a function member declaration, including the implicit `this` parameter, is valid in the entire function member. The ref-safe-scope of the fields of a `struct` type is function-member. A variable with ref-safe-scope of function-member is a valid referent only if the reference variable is declared in the same function member.
- ***caller-scope***: Member fields of a `class` type and `ref` parameters have ref-safe-scope of caller-scope. A variable with ref-safe-scope of caller-scope can be the referent of a reference return. A variable that can be the referent of a reference-return is ***safe-to-ref-return***.

> These values form a nesting relationship from narrowest (block) to widest (caller-scope).

Author: expand to included nested blocks.

- If `p` is a `ref`, or `in` parameter, its ref-safe-scope is the caller-scope. It is safe-to-ref-return. If `p` is an `in` parameter, it can't be returned as a writable `ref` but can be returned as `ref readonly`.
- If `p` is an `out` parameter, its ref-safe-scope is the function-member. It isn't safe-to-ref-return.
- Otherwise, if `p` is the `this` parameter of a struct type, its ref-safe-scope is the function-member.
- Otherwise, the parameter is a value parameter, and its ref-safe-scope is the function-member. It isn't safe-to-ref-return.

#### §ref-span-safety-field-reference Field ref safe scope

For a variable designating a reference to a field, `e.F`:

- If `e` is of a reference type, its ref-safe-scope is the caller-scope.
- Otherwise, if `e` is of a value type, its ref-safe-scope is the same as the ref-safe-scope of `e`.

#### §ref-span-safety-operators Operators

The conditional operator (§12.18), `c ? ref e1 : ref e2`, and reference assignment operator, `= ref e` (§12.21.1) have reference variables as operands and yield a reference variable. For those operators, the ref-safe-scope of the result is the narrowest scope among the ref-safe-scopes of all `ref` operands.

> #### §ref-span-safety-method-invocation Method invocation

Reviewer 6: Does this apply to other function member invocations as well, such as indexer invocations?

Author: Resolved at the meeting: generalize to all function invocations (local functions, property and indexer accessors, and delegates).

- A `ref struct` type may not be used as a type argument, or as an element type in a tuple type.
- A `ref struct` type may not be the declared type of a field, except that it may be the declared type of an instance field of another `ref struct`.
- A `ref struct` type may not be the element type of an array.
- A value of a `ref struct` type may not be boxed:
  - There is no conversion from a `ref struct` type to the type `object` or the type `System.ValueType`.
  - A `ref struct` type may not be declared to implement any interface
  - No instance method declared in `object` or in `System.ValueType` but not overridden in a `ref struct` type may be called with a receiver of that `ref struct` type.
  - No instance method of a `ref struct` type may be captured by method conversion to a delegate type.
- For a ref reassignment `ref e1 = ref e2`, the *ref_safe_scope* of `e2` must be at least as wide a scope as the *ref_safe_scope* of `e1`.
- For a ref return statement `return ref e1`, the *ref_safe_scope* of `e1` must be the *caller_scope*. In other words, `e1` must be *safe_to_return*.

> - A local function or anonymous function may not refer to a local or parameter of `ref struct` type declared in an enclosing scope.

Author: Let's put this one back here:

```suggestion
- For a method invocation if there is a `ref` or `out` argument of a `ref struct` type (including the receiver), with *ref_safe_scope* E1, then no argument (including the receiver) may have a narrower *ref_safe_scope* than E1.
- A local function or anonymous function may not refer to a local or parameter of `ref struct` type declared in an enclosing scope.
```

## §ref-span-safety Reference variables and returns

### §ref-span-safety-general General

A *reference variable* is a variable that refers to another variable, called the *referent*. Therefore, a *reference variable* does not store its value. Instead, a *reference variable* refers to its *referent*. Thus, the value of a reference variable is always the same as its referent variable. A *reference variable* is a local variable declared with the `ref` modifier.

> *Example:* The following example demonstrates a local *reference variable* whose *referent* is an element of an array:
>
> ```csharp
> public class C
> {
>
>     public void M()
>     {
>         int[] arr = new int[10];
>         // element is a reference variable that refers to arr[5]
>         ref int element = ref arr[5];
>         element += 5; // arr[5] has been incremented by 5
>     }     
> }
> ```
>
> *end example*

A *reference return* is the expression returned by reference from a method whose return type includes the `ref` or `ref readonly` modifiers (§15.6.1). The variable expression of the *reference return* is the *referent* of the *reference return*.

> > *Example:* The following example demonstrates a =*reference return* whose *referent* is an element of an array field:

Reviewer 2: Nit: remove `=` before *reference return*.

>
> ```csharp
> public class C
> {
>     private int[] arr = new int[10];
>
>     public readonly ref int M()
>     {
>         // element is a reference variable that refers to arr[5]
>         ref int element = ref arr[5];
>         return ref element; // return reference to arr[5];
>     }     
> }
> ```
>
> *end example*

> A `ref struct` may include `ref struct` or ref-like fields. A *ref-like field* refers to the same storage as its initializing expression. The initializing expression of the *ref_like field* is its *referent*. Unlike a reference field, a ref-like field may refer to a `struct` whose storage may be on the execution stack. Ref struct types include `Span<T>`, `ReadOnlySpan<T>`, and other types that include an unmanaged pointer as a member.

Reviewer 2: ref_like => ref-like

> > *Example:* The following example demonstrates a `ReadOnlySpan<T>` *reference variable* whose *referent* is the characters in a string. `ReadOnlySpan<T>` is a `ref struct`. One of its internal fields is a *ref-like field* that refers to the character array in the `string` referent. Therefore, `chars` is a *reference variable*:

Reviewer 2: Earlier we have: "A *reference variable* is a local variable declared with the `ref` modifier."

... but `chars` is apparently a reference variable, despite not being declared with the ref modifier. Perhaps the earlier sentence needs to include "or a variable whose type is a ref struct"?

Author: I reworded the text of the example. In this clause, I want to introduce that `ref struct` types contain reference variables. Therefore, they must not be copied outside of the ref safe scope of those reference variables.

>
> ```csharp
> string Reverse(string s)
> {
>     ReadOnlySpan<char> chars = s.AsSpan();
>     var reversed = new char[chars.Length];
>     for (int i = 0; i < chars.Length; i++)
>     {
>         reversed[chars.Length-i-1] = chars[i];
>     }
>     return new string(reversed);
> }
> ```
>
> *end example*

### §ref-span-safety-escape-scopes Ref safe scopes

All reference variables obey safety rules that ensure the scope of the reference variable is not greater than the *ref_safe_scope* of its referent.

> For any variable, the *ref_safe_scope* of that variable is the scope where a *variable_reference* (§9.5) to that variable is valid. The *referent* of a reference variable must have a *ref_safe_scope* that is at least as wide as the scope of the reference variable. A `ref struct` instance may not be copied by value beyond the value *ref_safe_scope* of the initializing expression(s) of any *ref-like fields*.

Reviewer 2: I'm not sure why we've got a mixture of dashes and underscores here. I *think* I'd expect dashes in descriptive names, but underscores in grammar elements. Would welcome @Nigel-Ecma's view here.

There are three valid *ref_safe_scope*s:

> - *block*: A *variable_reference* to a local variable declared in a block is valid in the block in which it is declared. The *ref_safe_scope* of a local variable is *block*. A local variable declared in a method has a *ref_safe_scope* of the *block* that defines the method. A variable declared in a block is a valid referent only if the reference variable is declared in the same block, or a nested block.

Reviewer 2: > A *variable_reference* to a local variable declared in a block is valid in the block in which it is declared.

Does it have to be definitely assigned? Presumably it's not valid prior to the declaration itself, i.e. I wouldn't *expect*:

```csharp
ref int x = ref y;
int y;
```

... to compile.

Author: Great catch. That doesn't compile. Updated the text.

> - *block*: A *variable_reference* to a local variable declared in a block is valid in the block in which it is declared. The *ref_safe_scope* of a local variable is *block*. A local variable declared in a method has a *ref_safe_scope* of the *block* that defines the method. A variable declared in a block is a valid referent only if the reference variable is declared in the same block, or a nested block.

Reviewer 2: > A local variable declared in a method has a *ref_safe_scope* of the *block* that defines the method.

What if the method doesn't have a block declaration, i.e. it's expression-bodied. I don't know whether an out parameter would count as "a local variable declared in a method"... this may not be important if you can't declare a reference variable other than in a block.

> - *block*: A *variable_reference* to a local variable declared in a block is valid in the block in which it is declared. The *ref_safe_scope* of a local variable is *block*. A local variable declared in a method has a *ref_safe_scope* of the *block* that defines the method. A variable declared in a block is a valid referent only if the reference variable is declared in the same block, or a nested block.

Reviewer 2: > A local variable declared in a method has a *ref_safe_scope* of the *block* that defines the method.

So something like this is presumably okay:

```csharp
public void M(string x)
{
    ReadOnlySpan<char> span1;
    if (x is not null)
    {
        ReadOnlySpan<char> span2 = x.AsSpan();
        span1 = span2;
    }
}
```

... because the ref_safe_scope of span2 is the whole of the method's block, not just the block of the if statement. (I haven't tested that - my machine is chugging along doing a large build at the moment.)

Author: Yes, that's valid.  See https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA+ABATARgLABQGAzAATakDCpA3oSeSqQLIAUGOADKQgJSF0CpYaQBKMAIYATAPIA7ADYBPAMoAHCXIA8YABYSoAPlIBnDXJwBuQiNIBLAGalWCeydJyIAFw8BXBQr8QiKCtrbi0vLK6po6+kam5likALw8AHQAgiYxcqy81sFhiZo4qSVyWIW2AL6EdQQ1QA

- *function_member*: A *variable_reference* to a value parameter on a function member declaration, including the implicit `this` parameter is valid in the entire function member. The *ref_safe_scope* of the fields of a `struct` type is *function_member*. A variable with *ref_safe_scope* of *function_member* is a valid referent only if the reference variable is declared in the same function member.

> - *caller_scope*: Member fields of a `class` type and `ref` parameters have *ref_safe_scope* of *caller_scope*. A variable with *ref_safe_scope* of *caller_scope* can be the referent of a reference return. A variable that can be the referent of a *reference-return* is *safe-to-ref-return*.

Reviewer 2: Do we need the last sentence here? This appears to be the only use of safe-to-ref-return.

Reviewer 2: Ah - that's because we later have *safe_to_return* by ref. So presumably this should be *safe_to_return* by ref? (I think we need a definition for this somewhere, and ideally making it a single italicised phrase - I personally prefer safe-to-ref-return, so maybe we should change everything else instead of this.)

Author: ~~I am confused by these two comments. Let's discuss at our meeting.~~

I figured it out from other comments. Replaced *safe-to-return* by ref with *safe-to-ref-return* throughout this PR.

These values form a nesting relationship from narrowest (*block*) to widest (*caller_scope*).

> > *Example*: The following code shows examples of the different *ref_safe_scope* values. The declarations show values of *ref_safe_scope* for a referent to be the initializing expression for a `ref` variable. The examples show the values of *ref_safe_scope* for a reference return:

Reviewer 2: Do we need the word "values" in this sentence? A ref_safe_scope feels a little less tangible than a normal value. Just *ref_safe_scope*s would work for me. (Likewise for the second sentence.)

>
> ```csharp
> public class C
> {
>     private int[] arr = { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9 };
> 
>     public ref int M1(ref int r1)
>     {
>         return ref r1; // r1 is safe to ref return
>     }
>
>     public ref int M2(int v1)
>     {
>         return ref v1; // error: v1 isn't safe to ref return
>     }
>
>     public ref int M3()
>     {
>         int v2 = 5;
> 
>         return ref arr[v2]; // arr[v2] is safe to ref return
>     }
> 
>     public void M4(int p) 
>     {
>         int v3 = 6;
>         ref int r2 = ref p; // scope of r2 is block, ref safe scope of p is method
>         ref int r3 = ref v3; // scope of r3 is block, ref safe scope of v3 is block
>         ref int r4 = ref arr[v3]; // scope of r4 is block, ref safe scope of arr[v3] is caller method
>     }
> }
> ```
>
> *end example.*

#### §ref-span-safety-locals Local variable ref safe scope

For a local variable `v`:

> - If `v` is a `ref` variable, its *ref_safe_scope* is taken from the *ref_safe_scope* of its initializing expression.

Reviewer 2: We've got "`ref` variable" here where earlier we've referred to "reference variable". If we mean "reference variable" here, please update it - importantly, `ReadOnlySpan<char> span = s.AsSpan();` is apparently a reference variable, but really doesn't look like a `ref` variable.

Reviewer 2: And could "is taken from" be "is the same as"? (That feels more precise to me.)

- Otherwise its *ref_safe_scope* is the scope in which it was declared.

#### §ref-span-safety-parameters Parameter ref safe scope

For a formal parameter `p`:

- If `p` is a `ref`, or `in` parameter, its *ref_safe_scope* is the *caller_scope*. It is *safe_to_return* by ref. If `p` is an `in` parameter, it can't be returned as a writable `ref` but can be returned as `ref readonly`.
- If `p` is an `out` parameter, its *ref_safe_scope* is the *function_member*. It isn't *safe_to_return* by ref.

> - Otherwise, if `p` is the `this` parameter of a struct type, its *ref_safe_scope* is the *function_member*.

Reviewer 2: And is it safe to ref return?

Author: Yes. ~~Leaving unresolved to discuss which term is preferred in our meeting.~~ Updated to *safe-to-ref-return*

> - Otherwise, the parameter is a value parameter, and it is *ref_safe_scope* is the *function_member*.

Reviewer 2: it is => its... and specify that it isn't safe to return by ref?

#### §ref-span-safety-field-reference Field ref safe scope

For a variable designating a reference to a field, `e.F`:

- If `e` is of a reference type, its *ref_safe_scope* is the *caller_scope*.

> - Otherwise, if `e` is of a value type, its *ref_safe_scope* is taken from the *ref_safe_scope* of `e`.

Reviewer 2: Again, I'd like a bit more clarity than "taken from" if possible.

#### §ref-span-safety-operators Operators

> The application of a user-defined operator is treated as a method invocation.

Reviewer 2: In meeting: can we remove this line entirely? (User-defined operators can't return by ref or have ref parameters.)

>
> *end example*

A *reference return* is the expression returned by reference from a method whose return type includes the `ref` or `ref readonly` modifiers (§15.6.1). The variable expression of the *reference return* is the *referent* of the *reference return*.

> *Example:* The following example demonstrates a =*reference return* whose *referent* is an element of an array field:
>
> ```csharp
> public class C
> {
>     private int[] arr = new int[10];
>
>     public readonly ref int M()
>     {
>         // element is a reference variable that refers to arr[5]
>         ref int element = ref arr[5];
>         return ref element; // return reference to arr[5];
>     }     
> }
> ```
>
> *end example*

> A `ref struct` may include `ref struct` or ref-like fields. A *ref-like field* refers to the same storage as its initializing expression. The initializing expression of the *ref_like field* is its *referent*. Unlike a reference field, a ref-like field may refer to a `struct` whose storage may be on the execution stack. Ref struct types include `Span<T>`, `ReadOnlySpan<T>`, and other types that include an unmanaged pointer as a member.

Author: I moved this description below the `return ref` description.

The one semantic change here says that `ref struct` types cannot be copied by value beyond the *ref_safe_scope* of its initializing expressions.

> *Example:* The following example demonstrates a `ReadOnlySpan<T>` *reference variable* whose *referent* is the characters in a string. `ReadOnlySpan<T>` is a `ref struct`. One of its internal fields is a *ref-like field* that refers to the character array in the `string` referent. Therefore, `chars` is a *reference variable*:
>
> ```csharp
> string Reverse(string s)
> {
>     ReadOnlySpan<char> chars = s.AsSpan();
>     var reversed = new char[chars.Length];
>     for (int i = 0; i < chars.Length; i++)
>     {
>         reversed[chars.Length-i-1] = chars[i];
>     }
>     return new string(reversed);
> }
> ```
>
> *end example*

### §ref-span-safety-escape-scopes Ref safe scopes

All reference variables obey safety rules that ensure the scope of the reference variable is not greater than the *ref_safe_scope* of its referent.

> For any variable, the *ref_safe_scope* of that variable is the scope where a *variable_reference* (§9.5) to that variable is valid. The *referent* of a reference variable must have a *ref_safe_scope* that is at least as wide as the scope of the reference variable. A `ref struct` instance may not be copied by value beyond the value *ref_safe_scope* of the initializing expression(s) of any *ref-like fields*.

Author: Here's the additional sentence.

Reviewer 2: I'm unclear what "of any ref-like fields" really means here. Any ref-like fields contributing to the ref struct instance? (How does the user know what's relevant here? Consider `s.AsSpan()` above - the user here doesn't know how that's implemented, or what fields are in `ReadOnlySpan<char>`. If `AsSpan` had another parameter - e.g. a `ref string x` - would that *necessarily* affect where the returned span could be copied, or would it depend on the implementation?)

(And is the word "value" in this sentence a typo? I'm not sure what that means either, if it's intentional.)

Author: Let's discuss some of this in our meeting. The wording is difficult until we get to C# 11, when `ref` fields are allowed. Admittedly, something like `AsSpan` is hard to deduce. Its implementation makes it more clear, in that it constructs the `Span` using a `ref char` to the first character. 

In these versions, `ref struct` types like `Span` and `ReadOnlySpan` use some attribute tricks to allow ref-like fields.

As for "value", I wanted to emphasize that returning a ref struct is returning a reference. Reworded.

Author: Removed from this PR, and added to the appropriate location on #601 

After thinking about Neal's points yesterday, I think the clearest way to spec this will be to have the ref safety rules in Variables. Those rules relating to the safe scope for a ref struct should be in the clause on ref structs. The rules apply, but trivially, to variables of other types. (Namely, ref structs are constrained to a safe scope. So are other variables, but for non-ref structs, the safe scope is calling method. Only ref structs have "interesting" constraints on where they can be copied.

>
> *end example*

For a value resulting from a method invocation `e1.M(e2, ...)`, its *safe_scope* from the narrowest of the following scopes:

- The *caller_scope*
- The *safe_scope* of all argument expressions (including the receiver)

A property invocation (either `get` or `set`) is treated as a method invocation of the underlying method by the above rules.

#### §ref-span-safety-a-value Values

A value's *ref_safe_scope* is the nearest enclosing scope.

> *Note:* This occurs in an invocation such as `M(ref d.Length)` where `d` is of type `dynamic`. It is also consistent with arguments corresponding to `in` parameters.

#### §ref-span-safety-stackalloc stackalloc

The *safe_scope* for a value `c` resulting from a `stackalloc` expression is the current method.

#### §ref-span-safety-constructor-invocations Constructor invocations

A `new` expression that invokes a constructor obeys the same rules as a method invocation (§ref-span-safety-method-invocation) that is considered to return the type being constructed.

> In addition, for a value `c` that is the result of a `new` expression, the *safe_scope* is no wider than the smallest of the *safe-scopes* of all arguments and operands of the object initializer expressions, recursively, if any initializer is present.

Reviewer 2: ```suggestion
In addition, for a value `c` that is the result of a `new` expression, the *safe_scope* is no wider than the smallest of the *safe_scope*s of all arguments and operands of the object initializer expressions, recursively, if any initializer is present.
```

- If `v` is a `ref` variable, its *ref_safe_scope* is taken from the *ref_safe_scope* of its initializing expression.
- Otherwise its *ref_safe_scope* is the scope in which it was declared.

The *safe_scope* of an expression that designates the use of a local variable is determined as follows:

- A value whose type is not a `ref struct` type is *safe_to_return* from the entire enclosing method. Its *safe_scope* is the *caller_scope*.
- If the variable is an iteration variable of a `foreach` loop, then the variable's *safe_scope* is the same as the *safe_scope* of the `foreach` loop's expression.
- A local of `ref struct` type which is uninitialized at the point of declaration is *safe_to_return* from the entire enclosing method. Its *safe_scope* is the *calling_method*.

> - Otherwise the variable's type is a `ref struct` type, and the variable's declaration has an initializer. The variable's *safe_scope* is the same as the *ref_safe_scope* of its initializer.

Author: These two rules need to be added to #601

- Otherwise, if `p` is the `this` parameter of a struct type, its *ref_safe_scope* is the *function_member*.
- Otherwise, the parameter is a value parameter, and it is *ref_safe_scope* is the *function_member*.

> For an expression that designates the value of a formal parameter `p`, its *safe_scope* is the calling method. This applies to the `this` parameter.

Author: I think this just works for `ref struct` types. But, check in #601

#### §ref-span-safety-field-reference Field ref safe scope

For a variable designating a reference to a field, `e.F`:

- If `e` is of a reference type, its *ref_safe_scope* is the *caller_scope*.
- Otherwise, if `e` is of a value type, its *ref_safe_scope* is taken from the *ref_safe_scope* of `e`.

> For an expression that designates a reference to a field, `e.F`, its a *safe_scope* is the same as the *safe_scope* of `e`.

Author: Add this rule to #601

#### §ref-span-safety-operators Operators

The application of a user-defined operator is treated as a method invocation.

> For an operator that yields a value, such as `e1 + e2` or `c ? e1 : e2`, the *safe_scope* of the result is the narrowest scope among the *safe_scopes* of the operands of the operator.  As a consequence, for a unary operator that yields an value, such as `+e`, the *safe_scope* of the result is the *safe_scope* of the operand.

Author: Add this rule to #601

For an operator that yields a variable, such as `c ? ref e1 : ref e2`:

- The *ref_safe_scope* of the result is the narrowest scope among the *ref_safe_scopes* of the operands of the operator.

> - The *safe_scope* of the operands must agree, and that is the *safe_scope* of the resulting variable.

Author: Add this rule to #601

>
> *end example*

For a value resulting from a method invocation `e1.M(e2, ...)`, its *safe_scope* from the narrowest of the following scopes:

- The *caller_scope*
- The *safe_scope* of all argument expressions (including the receiver)

> 

Author: This rule needs to be added to #601

> *Note:* This occurs in an invocation such as `M(ref d.Length)` where `d` is of type `dynamic`. It is also consistent with arguments corresponding to `in` parameters.

#### §ref-span-safety-stackalloc stackalloc

> The *safe_scope* for a value `c` resulting from a `stackalloc` expression is the current method.

Author: This needs to be added to either #601 or possibly #238

#### §ref-span-safety-constructor-invocations Constructor invocations

A `new` expression that invokes a constructor obeys the same rules as a method invocation (§ref-span-safety-method-invocation) that is considered to return the type being constructed.

> In addition, for a value `c` that is the result of a `new` expression, the *safe_scope* is no wider than the smallest of the *safe-scope*s of all arguments and operands of the object initializer expressions, recursively, if any initializer is present.

Author: This rule needs to be added to #601

#### §ref-span-safety-default-expressions Default expressions

> The *safe_scope* of a `default` expression is the *caller-scope*.

Author: Add this rule to #601

  - No instance method of a `ref struct` type may be captured by method conversion to a delegate type.
- For a ref reassignment `ref e1 = ref e2`, the *ref_safe_scope* of `e2` must be at least as wide a scope as the *ref_safe_scope* of `e1`.
- For a ref return statement `return ref e1`, the *ref_safe_scope* of `e1` must be the *caller_scope*. In other words, `e1` must be *safe_to_return*.
- For a return statement `return e1`, the *safe_scope* of `e1` must be the *caller_scope*. This is always true when `e1` is not a `ref struct`.
- For an assignment `e1 = e2`, if the type of `e1` is a `ref struct` type, then the *safe_scope* of `e2` must be at least as wide a scope as the *safe_scope* of `e1`.

> - For a method invocation if there is a `ref` or `out` argument of a `ref struct` type (including the receiver), with *safe_to_escape_scope* E1, then no argument (including the receiver) may have a narrower *safe_scope* than E1.

Author: These rules need to be added to #601
